lectures/monte_carlo.md

```diff
@@ -631,9 +642,69 @@ Notice that this version is faster than the one using a Python loop.
 Now let's try with larger $M$ to get a more accurate calculation.
 
 ```{code-cell} ipython3
-%%time 
+%%time
 compute_call_price(M=10_000_000)
 ```
 
 ```{solution-end}
 ```
+
+```{exercise}
+:label: monte_carlo_ex2
+
+Consider that a European call option may be written on an underlying with spot price of \$100 and a knockout barrier of \$120. This option behaves in every way like a vanilla European call, except if the spot price ever moves above \$120, the option "knocks out" and the contract is null and void. Note that the option does not reactivate if the spot price falls below \$120 again.
+Use the dynamics defined in {eq}`s_mc_dyms` to price the European call option.
+```
+
+```{solution-start} monte_carlo_ex2
+:class: dropdown
+```
+
+```{code-cell} ipython3
+μ  = 0.0001
+ρ  = 0.1
+ν  = 0.001
+S0 = 10
+h0 = 0
+K = 100
+n = 10
+β = 0.95
+bp = 120
+```
+
+```{code-cell} ipython3
+def compute_call_price_with_barrier(β=β,
+                                    μ=μ,
+                                    S0=S0,
+                                    h0=h0,
+                                    K=K,
+                                    n=n,
+                                    ρ=ρ,
+                                    ν=ν,
+                                    bp=bp,
+                                    M=10_000):
+    current_sum = 0.0
+    # For each sample path
+    for m in range(M):
+        s = np.log(S0)
+        h = h0
+        # Simulate forward in time
+        for t in range(n):
+            s_next = s + μ + np.exp(h) * randn()
+            h = ρ * h + ν * randn()
+            # Contract is closed as S_n > barrier price
+            if np.exp(s_next) > bp:
```

Comment:
I'm not sure if the logic is correct here.

If `np.exp(s_next) > bp` then we should not increase `current_sum`, since the payoff is zero.